lib/3scale/backend/stats/keys/generator/application.rb

```diff
@@ -0,0 +1,58 @@
+module ThreeScale
+  module Backend
+    module Stats
+      module Keys
+        module Generator
+          class Application
+            def initialize(service_context, index_limits)
+              self.service_context = service_context
+              self.index_limits = index_limits
+              self.applications = service_context.applications || []
+              self.metrics = service_context.metrics || []
+            end
+
+            def get_generator
+              Enumerator.new do |enum|
+                application_index_start, application_index_end = get_application_index_limits
+                metric_index_start, metric_index_end = get_metric_index_limits
+                self.applications[application_index_start..application_index_end].each_with_index do |app_id, app_range_idx|
+                  self.metrics[metric_index_start..metric_index_end].each_with_index do |metric_id, metric_range_idx|
+                    Timestamp.get_time_generator(self.service_context, self.index_limits, :app).each do |datetime_key, granularity_idx, ts|
+                      idx = Index.new(app: app_range_idx, metric: metric_range_idx, granularity: granularity_idx, ts: ts)
+                      stats_key = generate_stats_key(self.service_context.service_id, app_id, metric_id, datetime_key)
+                      enum << OpenStruct.new(key_index: idx, key_name: stats_key)
+                    end
```

Comment:
this is unreadable - things to do:
* make the `get_*_limits` return `Range` objects
* shorten the variable names, ie `app_range_idx` => `app_idx`, `metric_range_idx` => `m_idx, ...
* `self.<attribute>` is only useful for distinguishing a reader accessor from a local variable and a writer accessor call from a local variable assignment - none applies here, just use `<attribute>`
* I have not evaluated the use of `OpenStruct` yet, but if the fields are known beforehand you should use `Struct`
* this does not implement `Enumerable` (the `each` method) - it is usually enough to do just that and hide the Enumerator as an internal detail - are you perhaps looking for something like [enum_args](https://github.com/unleashed/enum_args)?
* unify this as much as possible with the metric equivalent

Reply:
* Regarding the get_*_limit returning Range objects: There are some limits where there is no concept of uniform/previsible succession. This is the case for example where the limits are timestamps. Another case is where the limits are the key limits, where the limits are not simple values but the combinations of other limits. A range accepts an initial value and a final value, and the implementation of the 'succ' method to calculate the intermediate values and <=> to compare them.
* About shortening the variable names, I'll do it :+1: 
* I'll change the self. by calls to the accessor method. This was a pending change indeed.
* I have created a Struct named IndexedKey instead of OpenStruct and will now be used